### PR TITLE
[inductor] test codegen with dynamic shapes

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96934

Adds new tests that check for patterns in generated C++/Triton code to
see if it's dynamic.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: .github/workflows/pull.yml
Comment (nkaretnikov) [inline .github/workflows/pull.yml:4]: **Remove this!** (to avoid running other jobs during testing)

File: test/inductor/expect/CodegenCpuTests.test_upsample_nearest1d_cpu.expect
Comment (nkaretnikov) [inline test/inductor/expect/CodegenCpuTests.test_upsample_nearest1d_cpu.expect:21]: Will need to update these paths. Left them for testing purposes.

Will also need to add all other files. Waiting for CI.

File: test/inductor/test_torchinductor_codegen.py
Comment (nkaretnikov) [inline test/inductor/test_torchinductor_codegen.py:37]: This is based on `check_model` and `check_model_cuda`

File: .ci/pytorch/test.sh
Comment (nkaretnikov) [inline .ci/pytorch/test.sh:259]: Add some diffing between static and dynamic outputs?

**Comment by nkaretnikov** [comment]
@ezyang Still WIP (see the comments above), seeking early feedback

File: .github/workflows/unstable.yml
Comment (nkaretnikov) [inline .github/workflows/unstable.yml:44]: **Add these two lines back!** (removed to avoid wasting compute during testing)

File: test/inductor/expect/CodegenCpuTests.test_avg_pool2d_backward4_cpu.expect
Comment (nkaretnikov) [inline test/inductor/expect/CodegenCpuTests.test_avg_pool2d_backward4_cpu.expect:18]: No CPP kernel.

File: test/inductor/expect/CodegenCpuTests.test_pixel_shuffle_channels_last_cpu.expect
Comment (nkaretnikov) [inline test/inductor/expect/CodegenCpuTests.test_pixel_shuffle_channels_last_cpu.expect:30]: Unrelated to this PR, but this is an interesting loop condition.

File: test/inductor/expect/CodegenCpuTests.test_std_cpu.expect
Comment (nkaretnikov) [inline test/inductor/expect/CodegenCpuTests.test_std_cpu.expect:128]: Same here and in a bunch of other places.

File: .lintrunner.toml
Comment (nkaretnikov) [inline .lintrunner.toml:373]: CPP codegen emits code with tabs and trailing spaces.

**Comment by ezyang** [comment]
This is too many expect tests. Which ones are you actually going to keep.

**Comment by nkaretnikov** [comment]
Okay, new plan. For CPP, I'll check `for` loops for `ks`. For triton, I'll check that `xnumel` is never assigned to an integer literal (or something similar).

File: test/inductor/test_torchinductor_codegen_dynamic_shapes.py
Comment (nkaretnikov) [inline test/inductor/test_torchinductor_codegen_dynamic_shapes.py:104]: Some of these are "partially dynamic": some loop might be dynamic, some not.

File: test/inductor/test_torchinductor_codegen_dynamic_shapes.py
Comment (nkaretnikov) [inline test/inductor/test_torchinductor_codegen_dynamic_shapes.py:154]: This includes code not having CPP code and CPP code that doesn't have any for loops.

**Comment by nkaretnikov** [comment]
@ezyang This is green. Thoughts on the current checks? CPP: `for` loop with integer literals, triton: `xnumel =` integer literal.

(If okay, I'll revert some WIP changes and enable the rest of CI.)

File: test/inductor/test_torchinductor_codegen_dynamic_shapes.py
Comment (ezyang) [inline test/inductor/test_torchinductor_codegen_dynamic_shapes.py:111]: Can we do xfails instead of skips?

**Comment by ezyang** [comment]
I'm willing to give this a shot. cc @ngimel for any other opinions.

File: test/inductor/test_torchinductor_codegen_dynamic_shapes.py
Comment (ngimel) [inline test/inductor/test_torchinductor_codegen_dynamic_shapes.py:41]: a comment on what this is testing would be nice

File: test/inductor/test_torchinductor_codegen_dynamic_shapes.py
Comment (ngimel) [inline test/inductor/test_torchinductor_codegen_dynamic_shapes.py:99]: if there's no static numel, there's still no guarantee that the code is dynamic?

Comment (nkaretnikov) [reply test/inductor/test_torchinductor_codegen_dynamic_shapes.py:99]: That's the first idea I had. I feel like checking for anything else might be error-prone. Open to suggestions.

File: test/inductor/test_torchinductor.py
Comment (nkaretnikov) [inline test/inductor/test_torchinductor.py:2140]: Here and elsewhere: changed the signature of `run_and_get_cpp_code` to match `run_and_get_triton_code`.

**Comment by nkaretnikov** [comment]
Status update: as of this commit, this only fails codegen tests on unstable, which means this shouldn't be executed outside of the unstable label (as intended). I'll add the missing codegen skips/xfails and this should be ready.

**Comment by nkaretnikov** [comment]
Note: before this is merged, I'd like to review the skip list here. I've just noticed we're not skipping `test_arange5` because it passes, but it doesn't use `common`. So should be in the skip list. There might be more like it.

File: .ci/pytorch/macos-test.sh
Comment (ezyang) [inline .ci/pytorch/macos-test.sh:49]: Shouldn't we default these tests as excluded, if we expect them to be failing most of the time?

**Comment by ezyang** [comment]
So what's our expected overall workflow once this lands? How are we keeping track of failures on the periodic flow?

**Comment by nkaretnikov** [comment]
@ezyang

> So what's our expected overall workflow once this lands? How are we keeping track of failures on the periodic flow?

Look at unstable jobs on hud occasionally. Run with ciflow/unstable on certain PRs. Fix issues once in a while.

However, the original plan was to check the enitre codegen output, which is how we decided on unstable. Now, we only look for small substrings. Shouldn't be too flaky. We also have a way to skip tests. I propose we enable this by default. Do you agree?

File: .ci/pytorch/macos-test.sh
Comment (nkaretnikov) [reply .ci/pytorch/macos-test.sh:49]: If we decide to go with unstable, OK. But see my reply below. I think we should just enable this as part of ordinary workflows -- no CI changes would be necessary, too.

File: test/inductor/test_torchinductor_codegen_dynamic_shapes.py
Comment (nkaretnikov) [inline test/inductor/test_torchinductor_codegen_dynamic_shapes.py:89]: Improved the dynamic var check a bit.

**Comment by nkaretnikov** [comment]
Pushed an update that removes CI changes, so that this runs as part of ordinary workflows. (My comments above explain why.)

**Comment by nkaretnikov** [comment]
@ezyang 3 minutes when I run locally:

```
pytest -rfxs test/inductor/test_torchinductor_codegen_dynamic_shapes.py  539.56s user 24.68s system 310% cpu 3:01.60 total
```

**Review by ezyang — APPROVED** [review]
OK, let's give a try, but yank it if it's causing people problems

**Pull request closed without merging**
